File: prototype/rails/src/test/actions/auth/claim_seller_identity_test.rb
require "identity_test_case"

module Auth
  class ClaimSellerIdentityTest < IdentityTestCase
    test "a first link for an address creates the seller" do
      seller = claim("newcomer@example.com")

      assert_equal "newcomer@example.com", seller.email
      assert_equal 1, Seller.count
    end

    test "a first link marks the address verified" do
      freeze_time do
        assert_equal Time.current, claim("newcomer@example.com").email_verified_at
      end
    end

    test "a later link returns the seller already holding the address" do
      existing = create_seller

      assert_equal existing, claim(existing.email)
      assert_equal 1, Seller.count
    end

    test "a later link leaves the original verification time alone" do
      existing = create_seller(email_verified_at: 3.days.ago)

      assert_equal existing.email_verified_at.to_i, claim(existing.email).email_verified_at.to_i
    end

    test "an address differing only in case reaches the same seller" do
      existing = create_seller(email: "artist@example.com")

      assert_equal existing, claim("ARTIST@Example.com")
    end

    private

    def claim(email)
      ClaimSellerIdentity.new.call(email: email)
    end
  end
end

File: prototype/node/src/app/actions/auth/claim-seller-identity.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { buildTestApp, TEST_INSTANT } from '../../test/build-test-app.ts'
import { claimSellerIdentity } from './claim-seller-identity.ts'
import { fixedClock } from '../../clock.ts'

test("a first link for an address creates the seller row and marks the address verified at the clock's now", async (t) => {
  const { db, clock, close } = await buildTestApp()
  t.after(close)

  const seller = await claimSellerIdentity({ db, clock }, 'artist@example.com')

  assert.equal(seller.email, 'artist@example.com')
  assert.equal(seller.emailVerifiedAt, TEST_INSTANT.toISOString())
})

test('a later link returns the seller already holding the address and creates no second row', async (t) => {
  const { db, clock, close } = await buildTestApp()
  t.after(close)

  const first = await claimSellerIdentity({ db, clock }, 'artist@example.com')
  const second = await claimSellerIdentity({ db, clock }, 'artist@example.com')

  assert.equal(second.id, first.id)
  const rows = await db.selectFrom('sellers').selectAll().where('email', '=', 'artist@example.com').execute()
  assert.equal(rows.length, 1)
})

test('a later link leaves the original verification time alone', async (t) => {
  const earlier = new Date('2026-08-01T00:00:00.000Z')
  const { db, close } = await buildTestApp({ clock: fixedClock(earlier) })
  t.after(close)

  await claimSellerIdentity({ db, clock: fixedClock(earlier) }, 'artist@example.com')
  const seller = await claimSellerIdentity({ db, clock: fixedClock(TEST_INSTANT) }, 'artist@example.com')

  assert.equal(seller.emailVerifiedAt, earlier.toISOString())
})

test('an address differing only in case reaches the same seller', async (t) => {
  const { db, clock, close } = await buildTestApp()
  t.after(close)

  const first = await claimSellerIdentity({ db, clock }, 'Artist@Example.com')
  const second = await claimSellerIdentity({ db, clock }, 'artist@example.com')

  assert.equal(second.id, first.id)
})
